### PR TITLE
Capture, receive, and RTP timestamp concept definitions & normative requirements for gUM/gDM

### PR DESCRIPTION
**Pull request by handellm**
Partly addresses https://github.com/w3c/webcodecs/pull/813#pullrequestreview-2169568794.

Defines capturetime in mediacapture-extensions.

***
<a href="https://pr-preview.s3.amazonaws.com/handellm/mediacapture-extensions/pull/156.html" title="Last updated on Oct 11, 2024, 2:04 PM UTC (fcafe1c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/mediacapture-extensions/156/14ff6a3...handellm:fcafe1c.html" title="Last updated on Oct 11, 2024, 2:04 PM UTC (fcafe1c)">Diff</a>

### CODE REVIEW

File: index.html
Comment (jan-ivar) [inline index.html:1162]: How does it help with AV sync if it's only on video frames?

File: index.html
Comment (jan-ivar) [inline index.html:1184]: What is the videoFrame.[timestamp](https://www.w3.org/TR/webcodecs/#dom-videoframe-timestamp) of these frames? Why do we need another one?

This is in milliseconds when the videoFrame.[timestamp](https://www.w3.org/TR/webcodecs/#dom-videoframe-timestamp) is in microseconds. That seems inconsistent.

File: index.html
Comment (jan-ivar) [inline index.html:1167]: What about {{Canvas/captureStream}}?

What if an HTMLMediaElement's srcObject is a MediaStream containing a video MediaStreamTrack from gUM() or gDM()? Does HTMLMediaElement/captureStream() generate a new timestamp or carry forward the original?

**Review by jan-ivar — CHANGES_REQUESTED** [review]
Why isn't the existing videoFrame.[timestamp](https://www.w3.org/TR/webcodecs/#dom-videoframe-timestamp) good enough for AV sync, especially when it's defined for both [audio and video](https://www.w3.org/TR/webcodecs/#dom-audiodata-timestamp), and has a higher resolution to boot?

File: index.html
Comment (handellm) [reply index.html:1184]: Hi, @jan-ivar - I was a little trigger happy in uploading this PR, I will upload changes soon.

File: index.html
Comment (jan-ivar) [reply index.html:1167]: Also, this doesn't say what user agents should do for other sources. Is it to leave always the member out? Or try to reconstitute it, if known somehow?

Comment (handellm) [reply index.html:1167]: The idea is to define in what way each source set the captureTime following recommendations in comments of https://github.com/w3c/webcodecs/pull/813, so this is just one of multiple PRs to make it fit together.

**Comment by jan-ivar** [comment]
The presentation timestamp seems a bit under-specified, but I gather it starts at 0 at the start of capture, and follows the "media timeline" which I suppose in theory can get out of sync (what happens to it with track.enabled = false?). So I can see a need here.

I just think it would be good to write down the differences and why we need a video capture timestamp.

**Comment by handellm** [comment]
41da924 is uploaded state prior to serving Youenn's request of extending with timestamp writing algorithms for local sources.

File: index.html
Comment (aboba) [inline index.html:1169]: Is captureTime defined as the time at which the first octet was captured (as it is in RTP)? Might be helpful to clarify.

File: index.html
Comment (aboba) [inline index.html:1174]: Since the rest of the text seems to assume a locally captured MST, introducing receiveTime raises questions.  If you are going to also cover remote tracks, then there will need to be text talking about which Wallclock is being used, whether there is an adjustment from the sender to receiver clock, etc.

File: index.html
Comment (aboba) [inline index.html:1196]: You'd need to add "of the sender" to the first sentence if you also will cover remote tracks.

**Review by aboba — COMMENTED** [review]
If you're going to cover remote tracks as well as local ones, then some clarifications are needed. But it might be simplest to focus on locally captured tracks.

File: index.html
Comment (handellm) [reply index.html]: `captureTime` is absolute which enables delay measurements in a video pipeline. Also, video capture OS APIs typically provide media sampling timestamps which I've measured to be sometimes tens of milliseconds before they're available in a capture sink. I've tried to update what the use of the timestamps are in non-normative text.

I don't know why `VideoFrame.timestamp` was specified in micros. DOMHighResTimestamp with resolution around 0.1ms seems fine.

File: index.html
Comment (handellm) [reply index.html]: Absolute audio capture timestamps are available via other web APIs right?
To try to answer the overarching question, presentation timestamps alone are insufficient if you want to do accurate AV sync on a local (think MediaRecorder) or on a remote receiver, especially in the presence of delays in the capture process, or insert VTGs in the capture pipeline to do video processing.
Is this answer enough or do you want to see further changes to clarify this in the spec text based on latest uploaded revision?

File: index.html
Comment (handellm) [reply index.html]: My gripe against the RTP def is that "octet" isn't accurate with how many capturers produce images (or if there are octets at all). 
I've extended the definition now, PTAL & resolve if happy.

File: index.html
Comment (handellm) [reply index.html:1179]: This piece of the text is defining unobservable "internal slots" and can't hold specifics or behavior; the thought is to do that for each source and sink instead - based on the TPAC session I thought there was consensus for this way?

Regarding wallclock, this is defined for the concrete VideoFrameMetadata below this section.

File: index.html
Comment (handellm) [reply index.html:1204]: Resolving for now due to other ongoing discussion in https://github.com/w3c/mediacapture-extensions/pull/156#pullrequestreview-2334237029.

**Review by henbos — APPROVED** [review]

File: index.html
Comment (aboba) [inline index.html:1221]: Is it the intent to reference this algorithm from another specification (e.g. mediacapture-transform)?

**Review by aboba — APPROVED** [review]

File: index.html
Comment (handellm) [reply index.html:1221]: Yes exactly!

Comment (dontcallmedom) [inline index.html:1221]:
```suggestion
      The <dfn class="abstract-op">Initialize Video Frame Timestamps From Internal MediaStreamTrack Video Frame</dfn>
```
(this marks the definition as being the name of an algorithm, and has it exported automatically)

File: index.html
Comment (dontcallmedom) [inline index.html:1229]:
```suggestion
      The <dfn class="abstract-op">Copy Video Frame Timestamps To Internal MediaStreamTrack Video Frame</dfn>
```

File: index.html
Comment (dontcallmedom) [inline index.html:1223]:
```suggestion
      <ol class=algorithm>
```
helps identify the steps as part of an algorithm

File: index.html
Comment (dontcallmedom) [inline index.html:1231]:
```suggestion
      <ol class=algorithm>
```

File: index.html
Comment (youennf) [inline index.html:1215]:
```suggestion
            <p>The RTP timestamp of the corresponding encoded frame. It corresponds to [=RTP timestamp=] of
```

File: index.html
Comment (youennf) [inline index.html:1222]: This is editorial, but I would use something like:
`When the XYZ algorithm is invoked with |frame| as input, run the following steps:`

File: index.html
Comment (youennf) [inline index.html:1241]:
```suggestion
        The user agent MUST set the [=capture timestamp=] of each video frame that is sourced from
```

File: index.html
Comment (youennf) [inline index.html:1244]: At TPAC, I think there was consensus on having camera tracks timestamp == capture timestamp.
It does not seem to be the case here though I would guess we keep the idea that the same clock is used.

File: index.html
Comment (handellm) [reply index.html]: Done.

File: index.html
Comment (handellm) [reply index.html]: > Should we state that the first video frame of a {{MediaStreamTrack}} has a presentation timestamp of 0?

This - updated with clarification that the first frame's timestamp is 0.

> Does this mean that the first video frame of a cloned track will have a timestamp of 0, or will cloned tracks have the same video frame timestamps?

Having it start with 0 for cloned tracks is consistent with the definition as written here. Trying to specify 0 for the first frame from a source could be interpreted as the same timestamp sequence is replicated across all tabs - we don't want this for privacy reasons, right?

**Review by Djuffin — APPROVED** [review]

File: index.html
Comment (youennf) [reply index.html:1245]: As per https://jsfiddle.net/4yzmwnsL/, it seems both Chrome and Safari use the same `frame.timestamp` for cloned tracks, even though the second track was cloned 2 seconds after the first one. This seems to be in contradiction with this PR.

Instead, implementations seem to be aligned with the idea that `capture timestamp` and `presentation timestamp` are both set by the source, not by the track. 

I am also not totally clear of the difference between these two timestamps for local sources.
It seems that `the time the frame is appearing on the track` would be the`presentation timestamp`, and would be slightly greater than the `capture timestamp`.
If that is the case, we should refer to both `capture timestamp` and `presentation timestamp` in that section of the PR.

Comment (handellm) [reply index.html:1245]: Yeah that's right, I spoke to this on TPAC. Currently, Chrome emits absolute capture timestamps from gUM/gDM capture to the webcodecs `timestamp` slot in MSTP. We believed we could do this since the nature of the timestamp isn't really well specified in webcodecs. It was later found that there are web apps that depend on a 0-based timeline and we have heuristics to support both cases in MSTG. The purpose of this PR series is to clean this whole topic up and stop using 1 field for 2 things.

`presentation timestamp` is 0-based and increments by frame duration. We intend to change Chrome back to 0-based here after the PR series is landed.
`capture timestamp` is absolute & unobservable on MediaStreamTracks but there are behavior requirements stated for gDM/gUM in this PR. It becomes observable first in `VideoFrameMetadata` creation at the MSTP.

> I am also not totally clear of the difference between these two timestamps for local sources.

For gUM/gDM sources, assume frame sequence indices 0, 1, 2, i ... N. `presentation timestamp[i + 1] - presentation timestamp[i] = capture timestamp[i + 1] - capture timestamp[i] = VideoFrame.duration[i]` i.e. same clock rate, different origin.

Suggestions on how to make this clearer?

Comment (youennf) [reply index.html:1245]: > The purpose of this PR series is to clean this whole topic up and stop using 1 field for 2 things.

That is fine to me.

> `presentation timestamp` is 0-based and increments by frame duration.

This PR assumes that `0` is per track. Meaning that track and track.clone() would not have the same definition of `0`. This seems unnatural to me. I would tend to go with `0` as the time of the first frame emitted by the source.
For VTG, `0` would refer to the time the first video frame is enqueued.

For this PR, that would mean changing `which is relative to the first frame appearing on the track.` to `which is relative to the first frame emitted by the track's [[Source]].`

Or it could be `0` per track's sink, meaning that MediaStreamTrackProcessor would always provide its first frame to the web app with a timestamp of `0` (provided the web app reads the frames quickly enough).

> It was later found that there are web apps that depend on a 0-based timeline

Maybe these web apps (and the heuristics you mentioned) could tell us whether `0` should be per track, per track's sink or per track's source.

> For gUM/gDM sources, assume frame sequence indices 0, 1, 2, i ... N. `presentation timestamp[i + 1] - presentation timestamp[i] = capture timestamp[i + 1] - capture timestamp[i] = VideoFrame.duration[i]` i.e. same clock rate, different origin.

That makes sense to me.

> Suggestions on how to make this clearer?

I would add some wording stating that `presentation timestamp` and `capture timestamp` are using the same clock and have a constant offset.

Comment (guidou) [reply index.html:1245]: > > `presentation timestamp` is 0-based and increments by frame duration.
> 
> This PR assumes that `0` is per track. Meaning that track and track.clone() would not have the same definition of `0`. This seems unnatural to me. I would tend to go with `0` as the time of the first frame emitted by the source. For VTG, `0` would refer to the time the first video frame is enqueued.
> 
> For this PR, that would mean changing `which is relative to the first frame appearing on the track.` to `which is relative to the first frame emitted by the track's [[Source]].`
> 
> Or it could be `0` per track's sink, meaning that MediaStreamTrackProcessor would always provide its first frame to the web app with a timestamp of `0` (provided the web app reads the frames quickly enough).
> 

I think `0` per track source is what makes most sense. WDYT?

Comment (handellm) [reply index.html:1250]: > > The purpose of this PR series is to clean this whole topic up and stop using 1 field for 2 things.
> 
> That is fine to me.
> 
> > `presentation timestamp` is 0-based and increments by frame duration.
> 
> This PR assumes that `0` is per track. Meaning that track and track.clone() would not have the same definition of `0`. This seems unnatural to me. I would tend to go with `0` as the time of the first frame emitted by the source. For VTG, `0` would refer to the time the first video frame is enqueued.
> 
> For this PR, that would mean changing `which is relative to the first frame appearing on the track.` to `which is relative to the first frame emitted by the track's [[Source]].`

> I think 0 per track source is what makes most sense. WDYT?

Done in [a8fc811](https://github.com/w3c/mediacapture-extensions/pull/156/commits/a8fc811bcdc51a5c3c32932ff535454efc81c9d1).


> Or it could be `0` per track's sink, meaning that MediaStreamTrackProcessor would always provide its first frame to the web app with a timestamp of `0` (provided the web app reads the frames quickly enough).
> 
> > It was later found that there are web apps that depend on a 0-based timeline
> 
> Maybe these web apps (and the heuristics you mentioned) could tell us whether `0` should be per track, per track's sink or per track's source.

The usages we found would work great if MSTP exposes 0-based from creation.


> > Suggestions on how to make this clearer?
> 
> I would add some wording stating that `presentation timestamp` and `capture timestamp` are using the same clock and have a constant offset.

Done in [a8fc811](https://github.com/w3c/mediacapture-extensions/pull/156/commits/a8fc811bcdc51a5c3c32932ff535454efc81c9d1).

**Comment by youennf** [comment]
> I think `0` per track source is what makes most sense.

That seems indeed better compared to per track.

> The usages we found would work great if MSTP exposes 0-based from creation.

That is closer to per track's sink timestamps, but...
MSTP can be specified so that the presentation timestamp of the video frames it is exposing to web pages will be relative to the first video frame it is exposing.

If we do that change in MSTP, do we need `presentation timestamp` and `capture timestamp` to be different for local tracks?

**Comment by handellm** [comment]
> > I think `0` per track source is what makes most sense.
> 
> That seems indeed better compared to per track.
> 
> > The usages we found would work great if MSTP exposes 0-based from creation.
> 
> That is closer to per track's sink timestamps, but... MSTP can be specified so that the presentation timestamp of the video frames it is exposing to web pages will be relative to the first video frame it is exposing.

I was expressing myself sloppily, I meant what you wrote. So for this PR, we could prepare for that by adding an |offset| parameter to the "Initialize Video Frame Timestamps From Internal MediaStreamTrack Video Frame" algorithm, which is specified later from mediacapture-transform as "the `[=presentation timestamp=]` of the first frame". WDYT? 

> If we do that change in MSTP, do we need `presentation timestamp` and `capture timestamp` to be different for local tracks?

In general yes, it's not possible to recreate any accurate `capture timestamp` given `presentation timestamp` alone since the origin is unknown.

**Comment by youennf** [comment]
> we could prepare for that by adding an |offset|

We could consider that MSTP is receiving a VideoFrame object and use https://w3c.github.io/webcodecs/#videoframe-initialize-frame-from-other-frame to create a new VideoFrame with an updated timestamp.

> > If we do that change in MSTP, do we need `presentation timestamp` and `capture timestamp` to be different for local tracks?
> 
> In general yes, it's not possible to recreate any accurate `capture timestamp` given `presentation timestamp` alone since the origin is unknown.

It seems that `presentation timestamp` will be exposed in a few places:
- Via MSTP where it would be updated by MSTP before exposure to JS.
- Via rvfc where it could also be updated in the same way (offset by the timestamp of the first video frame being rendered by the media element).

Based on this, the fact that the `presentation timestamp` of the first video frame of a track's source is `0` may not be observable to JavaScript, since the only `presentation timestamp`s exposed to JS would be computed by each sink and would be relative to that sink.

The only requirement could be that presentation timestamps have the same clock as capture timestamps for local sources. `presentation timestamp` == `capture timestamp` for local sources seems then sufficient and easier to understand.

**Comment by handellm** [comment]
> * Via MSTP where it would be updated by MSTP before exposure to JS.
> * Via rvfc where it could also be updated in the same way (offset by the timestamp of the first video frame being rendered by the media element).

Yes, `VideoFrame.timestamp` for the former and `mediaTime` for the latter (they're currently both described to be "presentation timestamp" or "media presentation timestamp (PTS)").

> The only requirement could be that presentation timestamps have the same clock as capture timestamps for local sources. `presentation timestamp` == `capture timestamp` for local sources seems then sufficient and easier to understand.

I buy requiring the same clock _rate_, but I don't see the requirement to require equality for the _offset_.

Given sinks that expose `presentation timestamp` subtract an |offset| being the first used `presentation timestamp` to produce a 0-based observable sequence it doesn't seem to me the actual origin of presentation timestamp in the UA matters and hence don't need to be strictly defined?

Wdyt we just add a note explaining this and that it's allowable to have the same clock (rate and offset) for `presentation timestamp` and `capture timestamp` as a special case?

**Comment by youennf** [comment]
> it doesn't seem to me the actual origin of presentation timestamp in the UA matters and hence don't need to be strictly defined?

Agreed.

> Wdyt we just add a note explaining this and that it's allowable to have the same clock (rate and offset) for `presentation timestamp` and `capture timestamp` as a special case?

Right, this seems editorial.
I would tend to define `presentation timestamp` equal to `capture timestamp` for simplicity/readability, but either is fine really.

**Comment by handellm** [comment]
@dontcallmedom I can't parse what respec complains about, ideas?

File: index.html
Comment (youennf) [inline index.html:1253]:
```suggestion
        the frame was captured.
```
The MUST is no longer observable.

**Review by youennf — APPROVED** [review]
LGTM once we say something of the [=presentation timestamp=] of local video capture, either same value as [=capture timestamp=] or with a fixed offset.
Plus the removal of a MUST we cannot really test.

**Comment by youennf** [comment]
> e [=presentation timestamp=] of local video capture, either same value as [=capture timestamp=]

Only for local tracks of course.

**Comment by jan-ivar** [comment]
Thanks for all the activity! Let me have a look. I'd also love for @padenot to look this over.

**Comment by henbos** [comment]
From WG meeting:
- @handellm to address @youennf 's comment
- @jan-ivar to read this PR

**Comment by padenot** [comment]
All good from my perspective, with comments from @youennf addressed. Please remember to put this upstream and not here, thanks.

**Comment by handellm** [comment]
@padenot thanks - can you just clarify what you meant by "upstream and not here"? We had one interpretation on the editors meeting but reading again I'm not sure that was right.

**Comment by happylinks** [comment]
Just want to share my appreciation for adding capture timestamp. This is exactly what we are missing for having accurate multi-stream recording in tella.tv 
Right now we start 2 mediarecorders and basically hope for the best in terms of sync between screen and camera, there is no way to know when the first frame was actually captured.
Seems like this will fix that problem, so thanks for working on it!

**Comment by handellm** [comment]
@padenot @jan-ivar - friendly ping!

**Review by guidou — APPROVED** [review]

**Review by jan-ivar — APPROVED** [review]
Thanks! I approve of overall direction. Happy to file followups if I find something.

**Pull request merged**
